server/src-lib/Hasura/Events/Lib.hs

```diff
@@ -199,107 +199,200 @@ processEvent
-
     getRetryAfterHeaderFromError (HStatus resp) = getRetryAfterHeaderFromResp resp
-    getRetryAfterHeaderFromError _ = Nothing
+    getRetryAfterHeaderFromError _              = Nothing
 
     getRetryAfterHeaderFromResp resp
       = let mHeader = find (\(HeaderConf name _) -> CI.mk name == retryAfterHeader) (hrsHeaders resp)
         in case mHeader of
              Just (HeaderConf _ (HVValue value)) -> Just value
              _                                   -> Nothing
-
     parseRetryHeader Nothing = Nothing
     parseRetryHeader (Just hValue)
       = let seconds = readMaybe $ T.unpack hValue
         in case seconds of
              Nothing  -> Nothing
              Just sec -> if sec > 0 then Just sec else Nothing
 
+encodeHeader :: EventHeaderInfo -> HTTP.Header
+encodeHeader (EventHeaderInfo hconf cache) =
+  let (HeaderConf name _) = hconf
+      ciname = CI.mk $ T.encodeUtf8 name
+      value = T.encodeUtf8 cache
+  in  (ciname, value)
+
+decodeHeader :: LogEnvHeaders -> [EventHeaderInfo] -> (HTTP.HeaderName, BS.ByteString) -> HeaderConf
+decodeHeader logenv headerInfos (hdrName, hdrVal)
+  = let name = decodeBS $ CI.original hdrName
+        getName ehi = let (HeaderConf name' _) = ehiHeaderConf ehi
+                      in name'
+        mehi = find (\hi -> getName hi == name) headerInfos
+    in case mehi of
+         Nothing -> HeaderConf name (HVValue (decodeBS hdrVal))
+         Just ehi -> if logenv
+                     then HeaderConf name (HVValue (ehiCachedValue ehi))
+                     else ehiHeaderConf ehi
+   where
+     decodeBS = TE.decodeUtf8With TE.lenientDecode
+
+addDefaultHeaders :: [HTTP.Header] -> [HTTP.Header]
+addDefaultHeaders hdrs = hdrs ++
+  [ (CI.mk "Content-Type", "application/json")
+  , (CI.mk "User-Agent", "hasura-graphql-engine/" <> T.encodeUtf8 currentVersion )
+  ]
+
+mkInvo :: EventPayload -> Int -> [HeaderConf] -> TBS.TByteString -> [HeaderConf] -> Invocation
+mkInvo ep status reqHeaders respBody respHeaders
+  = let resp = if isClientError status then mkClientErr respBody else mkResp status respBody respHeaders
+    in
+      Invocation
+      (epId ep)
+      status
+      (mkWebhookReq (toJSON ep) reqHeaders)
+      resp
+
+mkResp :: Int -> TBS.TByteString -> [HeaderConf] -> Response
+mkResp status payload headers =
+  let wr = WebhookResponse payload (mkMaybe headers) status
+  in ResponseType1 wr
+
+mkClientErr :: TBS.TByteString -> Response
+mkClientErr message =
+  let cerr = ClientError message
+  in ResponseType2 cerr
+
+mkWebhookReq :: Value -> [HeaderConf] -> WebhookRequest
+mkWebhookReq payload headers = WebhookRequest payload (mkMaybe headers) invocationVersion
+
+isClientError :: Int -> Bool
+isClientError status = status >= 1000
+
+mkMaybe :: [a] -> Maybe [a]
+mkMaybe [] = Nothing
+mkMaybe x  = Just x
+
+logQErr :: ( MonadReader r m, MonadIO m,  Has HLogger r) => QErr -> m ()
+logQErr err = do
+  (logger:: HLogger) <- asks getter
```

Comment:
Same. You don't need these annotations.

Reply:
In some other context, these variables were not resolving to correct type without annotation. But they seem to be fine now.